src/lib/storage/lz4/lz4_encoder.hpp

```diff
@@ -122,48 +137,348 @@ class LZ4Encoder : public SegmentEncoder<LZ4Encoder> {
+   * @param max_dictionary_size_estimate The original estimate for the maximum dictionary size
+   * @return The trained dictionary, or in the case of failure, an empty vector.
+   */
+  pmr_vector<char> _train_string_dictionary_padded(const pmr_vector<char>& values,
+                                                   const pmr_vector<size_t>& sample_sizes,
+                                                   const size_t max_dictionary_size_estimate) {
+    pmr_vector<char> dictionary;
+    size_t dictionary_size;
+    size_t max_dictionary_size = max_dictionary_size_estimate;
+
+    /**
+     * Work on copies of the input data and sample sizes. These vectors will increase in size by repeatedly appending
+     * the original input data and larger sample sizes.
+     */
+    auto values_copy = pmr_vector<char>{values, values.get_allocator()};
+    auto sample_sizes_copy = pmr_vector<size_t>{sample_sizes, sample_sizes.get_allocator()};
+
+    do {
+      /**
+       * This method is only called when the dictionary training with the input data failed. Therefore, we start by
+       * increasing the input data linearly.
+       */
+      dictionary = pmr_vector<char>{values.get_allocator()};
+      max_dictionary_size = std::min(2u * max_dictionary_size, _maximum_dictionary_size);
+      dictionary.resize(max_dictionary_size);
+      _increase_dictionary_input_data(values_copy, sample_sizes_copy, values.size());
+
+      dictionary_size =
+          ZDICT_trainFromBuffer(dictionary.data(), max_dictionary_size, values_copy.data(), sample_sizes_copy.data(),
+                                static_cast<unsigned>(sample_sizes_copy.size()));
+    } while (ZDICT_isError(dictionary_size) && max_dictionary_size < _maximum_dictionary_size &&
+             values_copy.size() < _maximum_value_size);
+
+    /**
+     * If the generation still failed with increased input data, then compress without a dictionary (the compression
+     * ratio will suffer).
+     */
+    if (ZDICT_isError(dictionary_size)) {
+      return pmr_vector<char>{};
+    }
+
+    DebugAssert(dictionary_size <= max_dictionary_size,
+                "Trained ZSTD dictionary in LZ4 compression is larger than "
+                "the memory allocated for it.");
+    dictionary.resize(dictionary_size);
+    dictionary.shrink_to_fit();
+
+    return dictionary;
+  }
+
+  /**
+   * Increase the dictionary input data by appending the original input data (linear increase). This only happens if
+   * the input data is too small for zstd to successfully train a dictionary.
+   *
+   * The sample sizes are not copied directly. First, we add a sample size of the whole input vector (somehow this is
+   * important for zstd to successfully train a dictionary).
+   * Afterwards, we again add the whole range of the input data as sample sizes of size 10 (and the last sample size
+   * varies according to the size of the input data).
```

Comment:
tbh, I can't follow the compression magic that you are doing here. From a code and documentation perspective, it looks good. If I worked my way deeper into the lz7 lib, this would probably make more sense to me. @Bouncner: I'd appreciate a check of the algorithms used here, if you can.

Reply:
We removed this part since Martin found out that it segfaults for a large amount of input data. Now, we just don't use a dictionary if the training fails.